src/runtime/swg-client-event-manager.js

```diff
@@ -0,0 +1,312 @@
+    this.isFromUserAction_ = value;
+  }
+
+  /**An optional parameter bag that can be included to provide additional
+   * details about some events.
+   * @return {Object}
+   */
+  getAdditionalParameters() {
+    return this.additionalParameters_;
+  }
+
+  /**An optional parameter bag that can be included to provide additional
+   * details about some events.  If you want to add or override a property,
+   * use the get method to get a reference to the existing object.
+   * @param {Object} value
+   */
+  setAdditionalParameters(value) {
+    assert(SwgClientEvent.isValidAdditionalParameters_(value),
+        'Invalid additionalParameters in SwgClientEvent (' + value + ')');
+    this.additionalParameters_ = value;
+  }
+
+  /**Returns true if the object is valid.  Otherwise it throws an error if you
+   * pass true and returns false if you don't.
+   * @param {boolean} createError
+   */
+  isValid(createError) {
+    if (!SwgClientEvent.isValidType_(this.eventType_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid event type in SwgClientEvent (' +
+          this.eventType_ + ')');
+    }
+    if (!SwgClientEvent.isValidOriginator_(this.eventOriginator_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid event originator in SwgClientEvent (' +
+          this.eventOriginator_ + ')');
+    }
+    if (!SwgClientEvent.isValidIsFromUserAction_(this.isFromUserAction_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid isFromUserAction in SwgClientEvent (' +
+          this.isFromUserAction_ + ')');
+    }
+    if (!SwgClientEvent.isValidAdditionalParameters_(
+        this.additionalParameters_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid additional parameters in SwgClientEvent (' +
+          this.additionalParameters_ + ')');
+    }
+    return true;
+  }
+}
+
+export class SwgClientEventManager {
+
+  /**True if the value is a function
+   * @param {*} value
+   * @private
+   */
+  static isFunction_(value) {
+    return typeof value === 'function';
+  }
+
+  constructor() {
+    /** @type {!CallBackArray} @private */
+    this.listeners_ = [];
+
+    /** @type {!CallBackArray} @private */
+    this.filterers_ = [];
+  }
+
+  /**Remove all existing filterers and listeners
+   */
+  clear() {
+    this.listeners_ = [];
+    this.filterers_ = [];
+  }
+
+  /**
+   * Ensures the callback function is notified anytime one of the passed
+   * events occurs unless a filterer returns false.
+   * @param {!SwgEventManagerListener} callback
+   * @public
+   */
+  addListener(callback) {
+    assert(SwgClientEventManager.isFunction_(callback),
+        'Invalid callback in SwgClientEventManager.addListener');
+    this.listeners_.push(callback);
+  }
+
+  /**
+   * Filterers are called before listeners, if it returns
```

Comment:
Register a filterer for events so that depending on the
result of the filtering operation, the event may be dropped
before calling the listeners registered.
@param {!SwGEventFilterer} callback

Reply:
I've taken some of your suggestion, let me know if you want further refinement.